core/process.py

```diff
@@ -1499,7 +1513,21 @@ def WhenDone(self, pid):
         if pid == self.last_stopped_pid:
             self.last_stopped_pid = -1
 
-        mylib.dict_erase(self.jobs, pid)
+        assert pid in self.child_procs
+        proc = self.child_procs[pid]
+        pipe = proc.parent_pipeline
+        if pipe:
+            if pipe.AllDone():
```

Comment:
I wonder if we should put the print statements in `Process:WhenDone()` and `Pipeline::WhenDone()`

We will have to change them to print the actual code to match POSIX/bash, and that will probably be easier there -- i.e. they will have a connection the original source code, not just the PID

I probably wasn't thinking about that when I initially added this

(I also want to rename this to WhenJobDone() -- I realized it is confusing to have methods of the same name on different classes!  It makes searching harder)

Reply:
I think we can actually remove `WhenDone()` from `JobList` entirely and replace it with `RemoveJob()`. All the logic other than updating `self.jobs` can live in the subclass' `WhenDone()`s